File: ui/backend/routes/mesh_prism_layers.py
"""POST /api/import/{case_id}/mesh/prism-layers — DEC-V61-137 (N2.3).

Engineer-driven boundary-layer prism injection on top of the existing
polyMesh from the gmsh stage. Adds a NEW mutating route that must be
registered in V132 ``MUTATING_ROUTES`` + behavioral contract test.
"""
from __future__ import annotations

from fastapi import APIRouter, HTTPException

from ui.backend.schemas.mesh_prism_layers import (
    MeshPrismLayersRequest,
    PrismFailingCheck,
    PrismLayersRejection,
    PrismLayersSuccessResponse,
    PrismLayersSummary,
)
from ui.backend.services.meshing_snappy.pipeline import (
    PrismLayersPipelineError,
    PrismLayersResult,
    apply_prism_layers,
)


router = APIRouter()


# Map pipeline failing_check enum → HTTP status. case_not_found is
# 404; polyMesh_not_ready / patch_not_found / snappy_diverged /
# snappy_addlayers_did_not_converge are all 422 (engineer-supplied
# config / state cannot run); snappy_container_failed is 502.
_STATUS_FOR_FAILING_CHECK: dict[PrismFailingCheck, int] = {
    "case_not_found": 404,
    "polyMesh_not_ready": 422,
    "patch_not_found": 422,
    "snappy_diverged": 422,
    "snappy_addlayers_did_not_converge": 422,
    "snappy_container_failed": 502,
}


def _result_to_response(result: PrismLayersResult) -> PrismLayersSuccessResponse:
    return PrismLayersSuccessResponse(
        case_id=result.case_id,
        prism_summary=PrismLayersSummary(
            cell_count=0,  # filled at the route layer if needed; v0
                          # leaves the raw polyMesh introspection to
                          # the V126/V127 mesh-quality endpoint, which
                          # the frontend re-fetches after this run.
            face_count=0,
            layers_added=result.layers_added,
            coverage_fraction=result.coverage_fraction,
            polyMesh_path=str(result.polyMesh_path),
            log_path=str(result.log_path),
            generation_time_s=result.generation_time_s,
        ),
    )


@router.post(
    "/import/{case_id}/mesh/prism-layers",
    response_model=PrismLayersSuccessResponse,
)
def mesh_prism_layers_route(
    case_id: str,
    request: MeshPrismLayersRequest,
) -> PrismLayersSuccessResponse:
    """Apply snappyHexMesh addLayers to the existing polyMesh.

    Synchronous on the FastAPI threadpool — snappyHexMesh runs
    O(seconds-minutes), same threadpool discipline as the gmsh
    route (V096 R5 P3 reference).
    """
    try:
        result = apply_prism_layers(
            case_id,
            patches=list(request.patches),
        )
    except PrismLayersPipelineError as exc:
        rejection = PrismLayersRejection(
            reason=str(exc),
            failing_check=exc.failing_check,
        )
        raise HTTPException(
            status_code=_STATUS_FOR_FAILING_CHECK.get(exc.failing_check, 400),
            detail=rejection.model_dump(),
        ) from exc

    return _result_to_response(result)

File: ui/frontend/src/types/mesh_prism_layers.ts
// Mirrors ui/backend/schemas/mesh_prism_layers.py — wire types only.
// DEC-V61-137 (N2.3): snappyHexMesh addLayers stage.

export const PRISM_EXPANSION_RATIO_MIN = 1.0;
export const PRISM_EXPANSION_RATIO_MAX = 2.0;
export const PRISM_MAX_LAYER_COUNT = 20;

// Per-patch prism configuration.
export interface PatchPrismConfig {
  patch: string;
  first_cell_height: number;
  expansion_ratio: number;
  num_layers: number;
}

// Request body for POST /api/import/{case_id}/mesh/prism-layers.
// v0 ships with `patches: list[PatchPrismConfig]` constrained to
// length 1; multi-patch is N2.3-extend.
export interface MeshPrismLayersRequest {
  patches: PatchPrismConfig[];
}

export type PrismFailingCheck =
  | "case_not_found"
  | "polyMesh_not_ready"
  | "patch_not_found"
  | "snappy_diverged"
  | "snappy_addlayers_did_not_converge"
  | "snappy_container_failed";

export interface PrismLayersSummary {
  cell_count: number;
  face_count: number;
  layers_added: number;
  coverage_fraction: number | null;
  polyMesh_path: string;
  log_path: string;
  generation_time_s: number;
}

export interface PrismLayersSuccessResponse {
  case_id: string;
  prism_summary: PrismLayersSummary;
}

export interface PrismLayersRejectionDetail {
  reason: string;
  failing_check: PrismFailingCheck | string;
}
